indispensable/src/commonMain/kotlin/at/asitplus/signum/indispensable/asn1/encoding/Asn1Decoding.kt

```diff
@@ -0,0 +1,401 @@
+@Throws(Asn1Exception::class)
+fun Asn1Primitive.decodeToULong() = runRethrowing { decode(Asn1Element.Tag.INT) { ULong.decodeFromAsn1ContentBytes(it) } }
+
+/** Exception-free version of [decodeToULong] */
+inline fun Asn1Primitive.decodeToULongOrNull() = runCatching { decodeToULong() }.getOrNull()
+
+/** Decode the [Asn1Primitive] as a [BigInteger]
+ * @throws [Asn1Exception] on invalid input */
+@Throws(Asn1Exception::class)
+fun Asn1Primitive.decodeToBigInteger() =
+    runRethrowing { decode(Asn1Element.Tag.INT) { BigInteger.decodeFromAsn1ContentBytes(it) } }
+
+/** Exception-free version of [decodeToBigInteger] */
+inline fun Asn1Primitive.decodeToBigIntegerOrNull() = runCatching { decodeToBigInteger() }.getOrNull()
+
+/**
+ * transforms this [Asn1Primitive] into an [Asn1String] subtype based on its tag
+ *
+ * @throws [Asn1Exception] all sorts of exceptions on invalid input
+ */
+@Throws(Asn1Exception::class)
+fun Asn1Primitive.asAsn1String(): Asn1String = runRethrowing {
+    when (tag.tagValue) {
+        UTF8_STRING.toULong() -> Asn1String.UTF8(String.decodeFromAsn1ContentBytes(content))
+        UNIVERSAL_STRING.toULong() -> Asn1String.Universal(String.decodeFromAsn1ContentBytes(content))
+        IA5_STRING.toULong() -> Asn1String.IA5(String.decodeFromAsn1ContentBytes(content))
+        BMP_STRING.toULong() -> Asn1String.BMP(String.decodeFromAsn1ContentBytes(content))
+        T61_STRING.toULong() -> Asn1String.Teletex(String.decodeFromAsn1ContentBytes(content))
+        PRINTABLE_STRING.toULong() -> Asn1String.Printable(String.decodeFromAsn1ContentBytes(content))
+        NUMERIC_STRING.toULong() -> Asn1String.Numeric(String.decodeFromAsn1ContentBytes(content))
+        VISIBLE_STRING.toULong() -> Asn1String.Visible(String.decodeFromAsn1ContentBytes(content))
+        else -> TODO("Support other string tag $tag")
+    }
+}
+
+/**
+ * Decodes this [Asn1Primitive]'s content into a String.
+ * @throws [Asn1Exception] all sorts of exceptions on invalid input
+ */
+fun Asn1Primitive.decodeToString() = runRethrowing {asAsn1String().value}
+
+/** Exception-free version of [decodeToString] */
+fun Asn1Primitive.decodeToStringOrNull() = runCatching { decodeToString() }.getOrNull()
+
+
+
+/**
+ * decodes this [Asn1Primitive]'s content into an [Instant] if it is encoded as UTC TIME or GENERALIZED TIME
+ *
+ * @throws Asn1Exception on invalid input
+ */
+@Throws(Asn1Exception::class)
+fun Asn1Primitive.decodeToInstant() =
+    when (tag) {
+        Asn1Element.Tag.TIME_UTC -> decode(Asn1Element.Tag.TIME_UTC, Instant.Companion::decodeUtcTimeFromAsn1ContentBytes)
+        Asn1Element.Tag.TIME_GENERALIZED -> decode(
+            Asn1Element.Tag.TIME_GENERALIZED,
+            Instant.Companion::decodeGeneralizedTimeFromAsn1ContentBytes
+        )
+
+        else -> TODO("Support time tag $tag")
+    }
+
+/**
+ * Exception-free version of [decodeToInstant]
+ */
+fun Asn1Primitive.decodeToInstantOrNull() = catching { decodeToInstant() }.getOrNull()
```

Comment:
```suggestion
fun Asn1Primitive.decodeToInstantOrNull() = runCatching { decodeToInstant() }.getOrNull()
```

Reply:
actually, runCatching is never safe, as it will even catch fatal JVM errors, which can never be recovered from. so it will hide the fact that we're in an unpredictable state. same goes for all other suggestions like this one

Reply:
...wait, what?

who _designed_ this language????